x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/inference/RestStartTrainedModelDeploymentAction.java

```diff
@@ -71,57 +74,109 @@ protected RestChannelConsumer prepareRequest(RestRequest restRequest, NodeClient
+                AllocationStatus.State.fromString(
+                    restRequest.param(WAIT_FOR.getPreferredName(), StartTrainedModelDeploymentAction.DEFAULT_WAITFOR_STATE.toString())
+                ),
+                StartTrainedModelDeploymentAction.DEFAULT_WAITFOR_STATE
+            )
+        );
+
+        RestCompatibilityChecker.checkAndSetDeprecatedParam(
+            NUMBER_OF_ALLOCATIONS.getDeprecatedNames()[0],
+            NUMBER_OF_ALLOCATIONS.getPreferredName(),
+            RestApiVersion.V_8,
+            restRequest,
+            (r, s) -> (Integer) sameParamInQueryAndBody(
+                request.getNumberOfAllocations(),
+                r.paramAsInt(s, StartTrainedModelDeploymentAction.DEFAULT_NUM_ALLOCATIONS),
+                StartTrainedModelDeploymentAction.DEFAULT_NUM_ALLOCATIONS
+            ),
+            request::setNumberOfAllocations
+        );
+        RestCompatibilityChecker.checkAndSetDeprecatedParam(
+            THREADS_PER_ALLOCATION.getDeprecatedNames()[0],
+            THREADS_PER_ALLOCATION.getPreferredName(),
+            RestApiVersion.V_8,
+            restRequest,
+            (r, s) -> (Integer) sameParamInQueryAndBody(
+                request.getThreadsPerAllocation(),
+                r.paramAsInt(s, StartTrainedModelDeploymentAction.DEFAULT_NUM_THREADS),
+                StartTrainedModelDeploymentAction.DEFAULT_NUM_THREADS
+            ),
+            request::setThreadsPerAllocation
+        );
+        request.setQueueCapacity(
+            (Integer) sameParamInQueryAndBody(
+                request.getQueueCapacity(),
+                restRequest.paramAsInt(QUEUE_CAPACITY.getPreferredName(), StartTrainedModelDeploymentAction.DEFAULT_QUEUE_CAPACITY),
+                StartTrainedModelDeploymentAction.DEFAULT_QUEUE_CAPACITY
+            )
+        );
+
+        if (restRequest.hasParam(CACHE_SIZE.getPreferredName())) {
+            request.setCacheSize(
+                (ByteSizeValue) sameParamInQueryAndBody(
+                    request.getCacheSize(),
+                    ByteSizeValue.parseBytesSizeValue(restRequest.param(CACHE_SIZE.getPreferredName()), CACHE_SIZE.getPreferredName()),
+                    null
                 )
             );
+        } else if (defaultCacheSize != null && request.getCacheSize() == null) {
+            request.setCacheSize(defaultCacheSize);
         }
 
+        request.setPriority(
+            (String) sameParamInQueryAndBody(
+                request.getPriority().toString(),
+                restRequest.param(StartTrainedModelDeploymentAction.TaskParams.PRIORITY.getPreferredName()),
+                StartTrainedModelDeploymentAction.DEFAULT_PRIORITY.toString()
+            )
+        );
+
         return channel -> client.execute(StartTrainedModelDeploymentAction.INSTANCE, request, new RestToXContentListener<>(channel));
     }
+
+    private Object sameParamInQueryAndBody(Object bodyParam, Object queryParam, Object paramDefault) {
```

Comment:
Let me know what you think of this: How about we determine which value to use by passing either the already set value from the body (`request.get*()`) or if that's null then a default value in the `restRequest.param*()` call?

I think that'd look something like this:

```
        var queueCap = request.getQueueCapacity();
        request.setQueueCapacity(
                restRequest.paramAsInt(QUEUE_CAPACITY.getPreferredName(), Objects.requireNonNullElse(queueCap, StartTrainedModelDeploymentAction.DEFAULT_QUEUE_CAPACITY))
        );
```

That way the query parameter will override the body if it is specified, or we'll take the body value, or if it wasn't set we'll either take the default already set when initializing the `StartTrainedModelDeploymentAction.Request` or we'll take the default passed to `Objects.requireNonNullElse`.

If I overlooked the logic and we still need this method then here are a few suggestions:
Let's make it static and I think it can be generic so we can remove the need to cast the return value.

Reply:
I think the main difference from what you're proposing is that the sameParamInQueryAndBody throws an exception if the body does not match the query param. 

Making the function static and generic does make sense though, I can do do that I think.